### ISSUE TITLE
Run code quality tests only once

### ISSUE THREAD

**Issue by schlessera** [wp-cli/wp-cli#3931]
The code quality tests (right now, only PHPCS) should only be run once, as their results won't change from one environment to the next. This could save some processing time for the pipeline.

Once Travis properly supports it, we might even avoid additional tests, once the code quality checks have failed.

**Comment by danielbachhuber** [comment]
> The code quality tests (right now, only PHPCS) should only be run once, as their results won't change from one environment to the next. This could save some processing time for the pipeline.

However, `phpcs` running on every build (and failing early if it's going to fail) will prevent the test suite from being run unnecessarily. 

So, seems like there's pros and cons both ways.

**Comment by johnbillion** [comment]
The new [build stages](https://blog.travis-ci.com/2017-05-11-introducing-build-stages) feature on Travis could work for this.

**Comment by danielbachhuber** [comment]
@schlessera I'm leaning towards `wontfix` for this.

Given how long the test suite takes to run, I think it's a feature that PHPCS fails early if needed. Doing so prevents the test suite from taking up a bunch of time when it doesn't need to.

Do you have a strong opinion the other way, or is it alright if I close?

**Comment by schlessera** [comment]
@danielbachhuber If I understand correctly, adding `fast-finish` to the Travis configuration would allow for what you want above, but get rid of multiple runs of PHPCS:

> With fast finishing enabled, Travis CI will mark your build as finished as soon as one of two conditions are met: The only remaining jobs are allowed to fail, or a job has already failed. In these cases, the status of the build can already be determined, so there’s no need to wait around until the other jobs finish.

So, having the PHPCS only on the first entry in the matrix, with fast-finish set to true, would immediately fail the entire build if PHPCS fails.

**Comment by danielbachhuber** [comment]
> So, having the PHPCS only on the first entry in the matrix, with fast-finish set to true, would immediately fail the entire build if PHPCS fails.

Ah. This would be fine then.

**Comment by GaryJones** [comment]
Once #4402 is (presumably) merged, I can handle this.

Notes:

```yml
# .travis.yml

matrix:
  fast_finish: true
  include:
    - php: 7.1
      env: WP_VERSION=latest SNIFF=1
```

```sh
# test.sh

# Run CodeSniffer
if [[ $SNIFF == '1' ]]
then
	vendor/bin/phpcs
fi
```

**Comment by danielbachhuber** [comment]
I think [Build Stages](https://docs.travis-ci.com/user/build-stages/) is the more correct way of implementing this.

**Comment by GaryJones** [comment]
Stages are currently in beta - are you sure you want to rely on it (especially when my notes above should be the only changes needed)?

**Comment by danielbachhuber** [comment]
We rely on the cron feature, which is also in beta. I think stages are the most correct way of implementing this.
